File: src/ucm/erikkarl/GraphCLI.java
// GraphCLI.java
package ucm.erikkarl;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Interfaz por consola que permite modificar un grafo y consultar el orden topologico de sus nodos
 * o sus componentes conexas segun proceda. Tambien permite hacer esto ultimo con grafos especificados
 * por ficheros de texto que siguen un formato concreto o incluso crear un fichero del estilo.
 */
public final class GraphCLI {
    /**
     * Scanner que lee de la entrada estandar.
     */
    private static final Scanner stdin = new Scanner(System.in);
    /**
     * Grafo que se mantiene en memoria el cual puede modificar el usuario.
     */
    private static Graph<Integer> graph = new Graph<>();

    private GraphCLI() {
    }

    /**
     * Crea un fichero de test con grafos aleatorios.
     * Pide al usuario especificar el numero de grafos, el numero de nodos
     * de cada uno y el nombre de un fichero donde guardar los datos del test.
     */
    private static void createTest() {
        System.out.print("Number of random tests: ");
        int numTests = readNumber();

        System.out.print("Number of nodes per graph: ");
        int numNodes = readNumber();
        var testCreator = new RandomTestCasesCreator(numTests, numNodes);

        System.out.print("Name of file: ");
        var fileName = readFileName();

        testCreator.createTestFile(fileName);
        System.out.println("Test created in file " + fileName + '.');
    }

    public static void readCommand() {
        System.out.print(">>> ");
        String commandText = stdin.nextLine();
        var optionalCommand = Arrays.stream(Command.values()).filter(x -> x.toString().equals(commandText)).findFirst();
        if (optionalCommand.isPresent())
            optionalCommand.get().execute();
        else {
            System.err.println("Error: this command doesn't exist.");
            System.err.println("Write " + Command.HELP.toString() + " to see the commands available.");
        }
    }

    /**
     * Muestra una introduccion al programa.
     */
    public static void showIntroduction() {
        System.out.println("This program allows you to create a graph manually, read test cases from a file," +
                " and even create random test files.");
        System.out.println("Write \"" + Command.HELP + "\" for help.\n");
    }

    /**
     * Muestra los comandos disponibles y su uso.
     */
    private static void showHelp() {
        for (Command c : Command.values()) {
            System.out.println("> " + c.toString() + "\n" + c.getHelp() + "\n");
        }
    }

    /**
     * Aplica el algoritmo especificado por el enunciado del ejercicio 4 de la practica
     * sobre el grafo {@link #graph}, es decir, haya el orden topologico de sus nodos
     * si no tiene bucles, y en caso contrario sus componentes fuertemente conexas.
     */
    private static void solve() {
        var initTime = System.nanoTime();
        System.out.println(Exercise4.solve(graph));
        var finalTime = System.nanoTime();
        System.out.println("Elapsed time: " + ((finalTime - initTime) / 1000000.0) + " ms");
    }

    /**
     * Pide al usuario el nombre de un fichero de test y delega la lectura del fichero
     * y la ejecucion del test. Despues, imprime algunos datos sobre los resultados por
     * consola. Estos datos mas la solucion
     */
    private static void runTest() {
        System.out.print("Name of file: ");
        var fileName = readFileName();
        var results = TestCasesManager.runTest(fileName);
        if (results != null) {
            System.out.println("Total number of cases: " + results.getNumberOfCases());
            System.out.println("Medium elapsed time per graph: " + results.mediumElapsedTimePerCase() + " ms");
            System.out.println("Total elapsed time: " + results.getTotalElapsedTime() + " ms");
            System.out.println("Results saved to " + results.getFileResultsName());
        } else {
            var currentPath = System.getProperty("user.dir");
            System.err.println("File does not exist in " + currentPath);
        }
    }

    /**
     * Muestra una representacion del grafo {@link #graph} por consola usando el
     * siguiente formato para cada nodo del grafo.
     * <p>
     * <code>Node: nodo -> [lista de nodos adyacentes]</code>
     * <p>
     */
    private static void showGraph() {
        if (!graph.iterator().hasNext()) {
            // If there are no nodes in the graph, the iterator has no next element
            System.out.println("Graph is empty");
        } else {
            for (Map.Entry<Integer, Set<Integer>> entry : graph) {
                var node = entry.getKey();
                var adjacents = entry.getValue();
                System.out.println("Node: " + node + "\t -> " + adjacents);
            }
        }
    }

    /**
     * El grafo {@link #graph} pasa a estar vacio, sin nodos.
     */
    private static void resetGraph() {
        graph = new Graph<>();
    }

    /**
     * Lee el nombre de un fichero por consola. No tiene por que existir pues esta funcion
     * puede ser usada para leer el nombre de un fichero a crear.
     */
    private static String readFileName() {
        String line = "";
        boolean validInput = false;

        while (!validInput) {
            line = stdin.nextLine();
            if (!line.isBlank() && !line.contains(" "))
                validInput = true;
            else {
                System.err.println("Invalid input: file name shouldn't be blank nor have whitespaces.");
            }
        }
        return line.trim();
    }

    /**
     * Lee un solo entero por consola.
     */
    private static int readNumber() {
        String line;
        int number = 0;
        boolean validInput = false;

        while (!validInput) {
            line = stdin.nextLine();
            try {
                number = Integer.parseInt(line);
                validInput = true;

            } catch (NumberFormatException e) {
                System.err.println("Invalid input: write a number.");
            }
        }

        return number;
    }

    /**
     * Lee una lista de nodos por consola. Todos deben ser enteros separados por espacios.
     */
    private static List<Integer> readNodesList() {
        Function<String, List<Integer>> formatToNodesList = (String x) ->
                Arrays.stream(x.split(" ")).map(Integer::valueOf).collect(Collectors.toList());
        String line;
        List<Integer> nodes = new LinkedList<>();
        boolean validInput = false;

        while (!validInput) {
            line = stdin.nextLine();
            try {
                nodes = formatToNodesList.apply(line);
                validInput = true;
            } catch (NumberFormatException e) {
                System.err.println("Invalid input: write a list of numbers.");
            }
        }
        return nodes;
    }

    /**
     * Pide al usuario un nodo y una lista de nodos adyacentes y
     * modifica el grafo {@link #graph}.
     */
    private static void addEdges() {
        System.out.print("Node: ");
        var node = readNumber();
        System.out.print("Adjacent nodes: ");
        var nodes = readNodesList();
        graph.addEdges(node, nodes);
    }

    /**
     * Pide al usuario un nodo que añadir al grafo {@link #graph}.
     */
    private static void addNode() {
        System.out.print("Node: ");
        var node = readNumber();
        graph.addNode(node);
    }

    /**
     * Clase para representar un comando. Cada comando tiene un mensaje propio de ayuda que se puede
     * consultar con el comando "help".
     */
    private enum Command {
        ADD_NODE {
            public String getHelp() {
                return "Adds a node to the graph without any adjacent nodes.";
            }

            public void execute() {
                addNode();
            }
        },
        ADD_EDGES {
            public String getHelp() {
                return "Creates directed edges that go from a source node to other nodes. Neither the source" +
                        " nor the destiny nodes have to exist already in the graph. The destiny nodes must have a space" +
                        " between them.";
            }

            public void execute() {
                addEdges();
            }
        },
        RESET {
            public String getHelp() {
                return "Resets the graph. It will have no nodes.";
            }

            public void execute() {
                resetGraph();
            }
        },
        HELP {
            public String getHelp() {
                return "Shows this message.";
            }

            public void execute() {
                showHelp();
            }
        },
        SHOW {
            public String getHelp() {
                return "Shows the nodes and its adjacent nodes in the graph.";
            }

            public void execute() {
                showGraph();
            }
        },
        SOLVE {
            public String getHelp() {
                return "(Ejercicio 4) Si el grafo es aciclico, lista los vertices en orden" +
                        " topologico. Si es ciclico, lista sus componentes fuertemente" +
                        " conexas como conjuntos de vertices (Kosaraju).";
            }

            public void execute() {
                solve();
            }
        },
        TEST {
            public String getHelp() {
                return "Reads a file with the data for several graphs and uses 'solve' with" +
                        " all of them.";
            }

            public void execute() {
                runTest();
            }
        },
        CREATE_TEST {
            public String getHelp() {
                return "Creates a random test file. The user has to specify the number of cases," +
                        " the number of nodes per graph and the name of the file.";
            }

            public void execute() {
                createTest();
            }
        };

        public abstract String getHelp();
        public abstract void execute();

        @Override
        public final String toString() {
            return this.name().replace('_', ' ').toLowerCase();
        }
    }
}

File: src/ucm/erikkarl/GraphCLI.kt
package ucm.erikkarl

import ucm.erikkarl.TestCasesManager.nanoToMilliseconds
import java.io.File
import kotlin.reflect.KFunction0
import kotlin.system.measureNanoTime


object GraphCLI {
    private var graph: Graph<Int> = Graph()

    private enum class Command {
        ADD_NODE {
            override fun getHelp(): String {
                return "Adds a node to the graph without any adjacent nodes."
            }

            override fun execute() {
                addNode()
            }
        },
        ADD_EDGES {
            override fun getHelp(): String {
                return "Creates directed edges that go from a source node to other nodes. Neither the source" +
                        " nor the destiny nodes have to exist already in the graph. The destiny nodes must have a space" +
                        " between them."
            }

            override fun execute() {
                addEdges()
            }
        },
        RESET {
            override fun getHelp(): String {
                return "Resets the graph. It will have no nodes."
            }

            override fun execute() {
                resetGraph()
            }
        },
        HELP {
            override fun getHelp(): String {
                return "Shows this message."
            }

            override fun execute() {
                showHelp()
            }
        },
        SHOW {
            override fun getHelp(): String {
                return "Shows the nodes and its adjacent nodes in the graph."
            }

            override fun execute() {
                showGraph()
            }
        },
        SOLVE {
            override fun getHelp(): String {
                return "(Ejercicio 4) Si el grafo es aciclico, lista los vertices en orden" +
                        " topologico. Si es ciclico, lista sus componentes fuertemente" +
                        " conexas como conjuntos de vertices (Kosaraju)."
            }

            override fun execute() {
                solve()
            }
        },
        TEST {
            override fun getHelp(): String {
                return "Reads a file with the data for several graphs and uses 'solve' with" +
                        " all of them"
            }

            override fun execute() {
                runTest()
            }
        },
        CREATE_TEST {
            override fun getHelp(): String {
                return "Creates a random test file. The user has to specify the number of cases," +
                        " the number of nodes per graph and the name of the file."
            }

            override fun execute() {
                createTest()
            }
        };

        abstract fun getHelp(): String
        abstract fun execute()

        fun commName(): String {
            return this.toString().replace('_', ' ').toLowerCase()
        }
    }

    private fun createTest() {
        print("Number of random tests: ")
        val numTests = readNumber()
        print("Number of nodes per graph: ")
        val numNodes = readNumber()
        val testCreator = RandomTestCasesCreator(numTests, numNodes)
        print("Name of file: ")
        val fileName = readFileName()
        testCreator.createTests(fileName)
        println("Test created in file $fileName.")
    }

    fun readCommand() {
        print(">>> ")
        val commandText: String = readLine()!!
        val command: Command? = Command.values().asSequence().find { it.commName().equals(commandText) }
        command?.execute() ?: println("Wrong command. Write ${Command.HELP.commName()} to see the commands available.")
    }

    fun showIntroduction() {
        println("This program allows you to create a graph manually or read test cases from a file.")
        println("Please, refer to the commands below to start using this program.")
        println()
        showHelp()
    }

    private fun showHelp() {
        for (comm in Command.values()) {
            println("${comm.commName()}\n\t${comm.getHelp()}\n")
        }
    }

    private fun solve() {
        val elapsedTime = measureNanoTime { println(Exercise4.solve(graph)) }
        println("Elapsed time: ${elapsedTime / 1000000.0} ms")
    }

    private fun runTest() {
        print("Name of file: ")
        val fileName = readLine()

        if (fileName != null && File(fileName).exists()) {
            val testCaseResults = TestCasesManager.runTest(fileName)!!
            println("Total number of cases: ${testCaseResults.numberOfCases}")
            println("Medium elapsed time per graph: ${nanoToMilliseconds(testCaseResults.mediumElapsedTimePerCase())} ms")
            println("Total elapsed time: ${nanoToMilliseconds(testCaseResults.totalElapsedTime)} ms")
        } else {
            val currentPath = System.getProperty("user.dir")
            System.err.println("File does not exist in $currentPath.")
        }
    }

    private fun showGraph() {
        if (graph.toList().isEmpty()) {
            println("Graph is empty")
        } else {
            for ((node, adjacents) in graph) {
                println("Node: $node \t-> $adjacents")
            }
        }
    }

    private fun resetGraph() {
        graph = Graph()
    }

    private fun readFileName(): String {
        var line: String? = readLine()
        while (line == null || line.isBlank()) {
            System.err.println("Not a valid file name.")
            print("Try again: ")
            line = readLine()
        }
        return line
    }

    private fun readNumber(): Int {
        var line: String? = readLine()
        while (line == null || line.toIntOrNull() == null) {
            System.err.println("Not a number.")
            print(" Try again: ")
            line = readLine()
        }
        return line.toInt()
    }

    private fun readNodesList(): List<Int> {
        var line: String? = readLine()
        val formatToNodesList: (String) -> List<Int?> = { x -> x.split(' ').map { it.toIntOrNull() } }

        while (line == null || formatToNodesList(line).any { it == null }) {
            System.err.println("All nodes have to be numbers.")
            print(" Try again: ")
            line = readLine()
        }
        return formatToNodesList(line) as List<Int>
    }

    private fun addEdges() {
        print("Node: ")
        val node = readNumber()
        print("Adjacent nodes: ")
        val nodes = readNodesList()
        graph.addEdges(node, nodes)
    }

    private fun addNode() {
        print("Node: ")
        val node = readNumber()
        graph.addNode(node)
    }

}

typealias Component = Either<Int, List<Int>>